src/engine/loopingcontrol.cpp

```diff
@@ -367,157 +387,167 @@ void LoopingControl::hintReader(HintVector* pHintList) {
         // aren't that bad to make anyway.
-        if (m_iLoopStartSample >= 0) {
+        if (loopSamples.start >= 0) {
             loop_hint.priority = 2;
-            loop_hint.sample = m_iLoopStartSample;
+            loop_hint.sample = loopSamples.start;
             loop_hint.length = 0; // Let it issue the default length
             pHintList->append(loop_hint);
         }
-        if (m_iLoopEndSample >= 0) {
+        if (loopSamples.end >= 0) {
             loop_hint.priority = 10;
-            loop_hint.sample = m_iLoopEndSample;
+            loop_hint.sample = loopSamples.end;
             loop_hint.length = -1; // Let it issue the default (backwards) length
             pHintList->append(loop_hint);
         }
     } else {
-        if (m_iLoopStartSample >= 0) {
+        if (loopSamples.start >= 0) {
             loop_hint.priority = 10;
-            loop_hint.sample = m_iLoopStartSample;
+            loop_hint.sample = loopSamples.start;
             loop_hint.length = 0; // Let it issue the default length
             pHintList->append(loop_hint);
         }
     }
 }
 
 void LoopingControl::slotLoopIn(double val) {
-    if (!m_pTrack) {
+    if (!m_pTrack || val <= 0.0) {
         return;
     }
-    if (val) {
-        clearActiveBeatLoop();
 
-        // set loop-in position
-        int pos =
-                (m_pQuantizeEnabled->get() > 0.0 && m_pClosestBeat->get() != -1) ?
-                static_cast<int>(floor(m_pClosestBeat->get())) : m_iCurrentSample;
-
-        // If we're looping and the loop-in and out points are now so close
-        //  that the loop would be inaudible (which can happen easily with
-        //  quantize-to-beat enabled,) set the in point to the smallest
-        //  pre-defined beatloop size instead (when possible)
-        if (m_bLoopingEnabled &&
-            (m_iLoopEndSample - pos) < MINIMUM_AUDIBLE_LOOP_SIZE) {
-            pos = m_iLoopEndSample;
-            if (m_pQuantizeEnabled->get() > 0.0 && m_pBeats) {
-                // 1 would have just returned loop_in, so give 2 to get the beat
-                // following loop_in
-                int nextbeat = m_pBeats->findNthBeat(pos, 2);
-                pos -= (nextbeat - pos) * s_dBeatSizes[0];
-            }
-            else pos -= MINIMUM_AUDIBLE_LOOP_SIZE;
-        }
+    clearActiveBeatLoop();
 
-        if (pos != -1 && !even(pos)) {
-            pos--;
+    // set loop-in position
+    LoopSamples loopSamples = m_loopSamples.getValue();
+    double closestBeat = -1;
+    int pos = m_iCurrentSample;
+    if (m_pQuantizeEnabled->toBool()) {
+        closestBeat = m_pClosestBeat->get();
+        if (closestBeat != -1) {
+            pos = static_cast<int>(floor(closestBeat));
         }
+    }
 
-        m_iLoopStartSample = pos;
-        m_pCOLoopStartPosition->set(m_iLoopStartSample);
+    // Reset the loop out position if it is before the loop in so that loops
+    // cannot be inverted.
+    if (loopSamples.end != kNoTrigger &&
```

Comment:
This is still found by the MINIMUM_AUDIBLE_LOOP_SIZE check, because negative is also smaller.